velox/common/memory/HashStringAllocator.cpp

```diff
@@ -591,6 +591,77 @@ void HashStringAllocator::ensureAvailable(int32_t bytes, Position& position) {
   position = finishWrite(stream, 0).first;
 }
 
+inline bool HashStringAllocator::storeStringFast(
+    const char* bytes,
+    int32_t numBytes,
+    char* destination) {
+  if (numFree_ != 1 || free_[kNumFreeLists - 1].empty()) {
+    return false;
+  }
+  if (numBytes < kMinAlloc) {
+    numBytes = kMinAlloc;
+  }
+  auto& freeList = free_[kNumFreeLists - 1];
+  auto header = headerOf(freeList.next());
+  auto size = header->size();
+  if (numBytes > size) {
+    return false;
+  }
+  auto spaceTaken = numBytes + sizeof(Header);
+  if (spaceTaken + kMinAlloc > size) {
+    // Takes the whole entry.
+    --numFree_;
+    freeBytes_ -= header->size() + sizeof(Header);
+    removeFromFreeList(header);
+    if (auto* next = header->next()) {
+      next->clearPreviousFree();
+    }
+  } else if (size - spaceTaken >= kMaxAlloc) {
+    // The entry after allocation stays in the largest free list.
+    // The size at the end of the block is changed in place.
+    reinterpret_cast<int32_t*>(header->end())[-1] -= spaceTaken;
+    auto freeHeader =
+        new (header->begin() + numBytes) Header(header->size() - spaceTaken);
+    freeHeader->setFree();
+    header->clearFree();
+    memcpy(freeHeader->begin(), header->begin(), sizeof(CompactDoubleList));
+    freeList.updateNext(
+        reinterpret_cast<CompactDoubleList*>(freeHeader->begin()));
+    header->setSize(numBytes);
+    freeBytes_ -= spaceTaken;
+  } else {
+    // The remaining block is no longer in the last free list.
+    --numFree_;
+    freeBytes_ -= header->size() + sizeof(Header);
+    removeFromFreeList(header);
+    freeRestOfBlock(header, spaceTaken);
```

Comment:
Do we need to do `header->next()->clearPreviousFree()` here as well?

Reply:
@oerling Actually I don't think you need to separate this case with the first `if` branch.  Just adding `freeRestOfBlock` there, adjust the condition, and remove this branch should be good.

Reply:
I see it should be good as it is .  Still could be merged with the first branch though.